File: IPLS2OC10/F.py
n, m = [int(a) for a in input().split()]
x, y = [int(a) for a in input().split()]
x-=1
y-=1
adj = [[] for _ in range(n)]
for i in range(m):
    u,v,w = [int(a) for a in input().split()]
    u-=1
    v-=1
    adj[u].append((v,w))
    adj[v].append((u,w))
for i in range(n):
    adj[i].sort(key=lambda x: x[1], reverse=True)

taxi = []
for i in range(n):
    t,c = [int(a) for a in input().split()]
    taxi.append((t,c))

dp = [2**1000]*n
dp[x] = 0
marked1 = [False]*n
stk1 = [x]
while stk1:
    u1 = stk1.pop()
    if marked1[u1]:
        continue
    marked1[u1] = True

    # see where you can get in t_u1
    marked2 = [False]*n
    t_u1, c_u1 = taxi[u1]
    stk2 = [(u1,0)]
    while stk2:
        u2,d = stk2.pop()
        if marked2[u2]:
            continue
        marked2[u2] = True

        dp[u2] = min(dp[u2], dp[u1]+c_u1)
        for v2,w in adj[u2]:
            if not marked2[v2] and w+d <= t_u1:
                stk2.append((v2,w+d))

    for v1,w in adj[u1]:
        if not marked1[v1]:
            stk1.append(v1)
if dp[y] == 2**1000:
    print(-1)
else:
    print(dp[y])

File: IPLS2OC10/F.cpp
#include <iostream>
#include <stdio.h>
#include <stdlib.h>
#include <algorithm>
#include <cmath>
#include <climits>
#include <vector>
#include <queue>
#include <cstring>
#include <iterator>
#include <list>
#include <set>
#include <map>
#include <bitset>
using namespace std;

typedef long long LL;
typedef pair<LL,LL> PLL;

vector<PLL> adj[1010];
LL t[1010];
LL c[1010];
LL dp[1010];
bool marked1[1010];
bool marked2[1010];
LL infinity = LLONG_MAX;
int main()
{
    LL n, m, x, y;
    cin >> n >> m >> x >> y;
    x--; y--;

    for (int i = 0; i < m; i++) {
        LL u, v, w;
        cin >> u >> v >> w;
        u--; v--;
        adj[u].push_back(make_pair(v,w));
        adj[v].push_back(make_pair(u,w));
    }

    for (int i = 0; i < n; i++) {
        cin >> t[i] >> c[i];
    }

    for (int i = 0; i < n; i++) dp[i] = infinity;
    dp[x] = 0;
    memset(marked1, false, sizeof(marked1));
    vector<LL> stk1;
    stk1.push_back(x);
    while(!stk1.empty()) {
        LL u1 = stk1.back();
        stk1.pop_back();
        if (marked1[u1])
            continue;
        marked1[u1] = true;

        memset(marked2, false, sizeof(marked2));
        vector<PLL> stk2;
        stk2.push_back(make_pair(u1,0));
        while(!stk2.empty()) {
            PLL p2 = stk2.back();
            LL u2 = p2.first;
            LL d = p2.second;
            stk2.pop_back();
            if(marked2[u2])
                continue;
            marked2[u2] = true;

            dp[u2] = min(dp[u2], dp[u1]+c[u1]);
            for (PLL edge : adj[u2]) {
                LL v2 = edge.first;
                LL w = edge.second;
                if (!marked2[v2] && w+d <= t[u1])
                    stk2.push_back(make_pair(v2,w+d));
            }
        }

        for (auto edge : adj[u1]) {
            LL v1 = edge.first;
            LL w = edge.second;
            if (!marked1[v1])
                stk1.push_back(v1);
        }
    }

    if (dp[y] == infinity)
        cout << -1 << endl;
    else
        cout << dp[y] << endl;
}
